Start-stop file writer from Nicos

umithardal opened the pull request:
This PR adds the capability to start and stop a file writer instance from Nicos (Ymir) GUI. 

While the current code can be used without any problems, it is not finished. In particular,

- `print` calls should be exchanged with proper log calls,
- `sleep` calls should be exchanged with proper validations,
- There should be a validation upon stop.

UPDATE:

- `print` now removed and `session.log.info` is used
- `sleep` removed and a reusable generator is implemented
- It is decided that the client has no need to do any validation of the written file.

mattclarke commented on nicos_ess/devices/datasinks/file_writer.py:
```diff
@@ -0,0 +1,87 @@
+        # Check if the process is still running
+        # if self._mode == MASTER and not self.is_process_running():
+        #     self._setROParam('curstatus', (status.ERROR, 'Disconnected'))
+        # TODO: not urgent but make it work like status_handler.py
```
Can you create a ticket for this TODO then delete the commented out code please?
Just so we don't lose track of it :)

mattclarke commented on nicos_ess/ymir/setups/system.py:
```diff
@@ -45,33 +45,32 @@
-               data_topic='event_data', brokers=['172.30.242.20:9092'],
-               unit='evts', command_topic='hist_commands',
-               response_topic='hist_responses'),
+    # det=device('nicos_ess.devices.datasources.just_bin_it.JustBinItDetector',
```
We need to remember to uncomment det before merging

mattclarke commented on nicos_ess/ymir/setups/system.py:
```diff
@@ -45,33 +45,32 @@
-startupcode = '''
-SetDetectors(det)
-'''
+# startupcode = '''
```
We need to remember to uncomment this before merging too

mattclarke commented on nicos_ess/ymir/setups/system.py:
```diff
@@ -45,33 +45,32 @@
+    FileWriterParameters=device(
+        'nicos_ess.devices.datasinks.file_writer.FileWriterParameters',
+        description='File-writer parameters',
+        broker=['192.168.0.36:9092'],
```
We will need to change the broker address to the correct values for YMIR before merging/deploying.
172.30.242.20

mattclarke commented on nicos_ess/ymir/setups/system.py:
```diff
@@ -45,33 +45,32 @@
+    FileWriter=device(
+        'nicos_ess.devices.datasinks.file_writer.FileWriterStatus',
+        description='Status for file-writer',
+        broker=['192.168.0.36:9092'],
```
We will need to change the broker address to the correct values for YMIR before merging/deploying.
172.30.242.20

mattclarke commented on nicos_ess/ymir/setups/system.py:
```diff
@@ -45,33 +45,32 @@
+        description='File-writer parameters',
+        broker=['192.168.0.36:9092'],
+        command_topic='UTGARD_writerCommandStatus',
+        nexus_config_path='nicos_ess/ymir/commands/nexus_config.json',
```
We should set this device to lowlevel, so it doesn't appear in the devices list for the user.

umithardal replied:
Could you elaborate?

mattclarke replied:
No problem.
NICOS has a lowlevel flag which means devices don't appear unless Expert Mode is enabled. For example: if you enable expert mode in the GUI you will see conssink and daemonsink appear.

I think the parameters device is probably a device the user doesn't need to see.

mattclarke commented on nicos_ess/ymir/commands/file_writer.py:
```diff
@@ -0,0 +1,39 @@
+@usercommand
+def start_writing():
+    writer = StartFileWriter()
+    writer.start_job()
```
What happens if the user accidentally tries to start_writing again without calling stop_writing?
I think the FW will ignore the 2nd command but we need to make sure NICOS doesn't get confused.

umithardal replied:
This was tricky :) But handled!

mattclarke reviewed (commented):
Looking good (and it works!), just a few comments to take into account while you are finishing it off
Good job!

umithardal commented:
@mattclarke I have now implemented a generator `wait_until_true(conditions)` where conditions are boolean valued callables. It is now in a new directory called `utilities/managers.py` and can be reused in similar situations. I expect `managers.py` is going to be extended to has all the generators and context-managers we come with.

For logging I have coordinated with Jens and now implemented `session.log.info()` which works without a problem.  

I am planning to finish up everything here and then push to Gerrit, if that is OK?

mattclarke commented:
Sounds good to me.


From: AUC HARDAL <notifications@github.com>
Reply-To: ess-dmsc/nicos <reply@reply.github.com>
Date: Friday, 4 December 2020 at 11:38
To: ess-dmsc/nicos <nicos@noreply.github.com>
Cc: Matt Clarke <matt.clarke@ess.eu>, Mention <mention@noreply.github.com>
Subject: Re: [ess-dmsc/nicos] Start-stop file writer from Nicos (#39)


@mattclarke<https://github.com/mattclarke> I have now implemented a generator wait_until_true(conditions) where conditions are boolean valued callables. It is now in a new directory called utilities/managers.py and can be reused in similar situations. I expect managers.py is going to be extended to has all the generators and context-managers we come with.

For logging I have coordinated with Jens and now implemented session.log.info() which works without a problem.

I am planning to finish up everything here and then push to Gerrit, if that is OK?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/ess-dmsc/nicos/pull/39#issuecomment-738709158>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAXELLY2TQHUUAKOCMNPAU3STC3YVANCNFSM4UHTKG5Q>.

umithardal commented:
Let us polish the code before moving it. Please review when you are available :) @mattclarke

mattclarke commented:
You’ve addressed my comments so, unless there is something specific you want me to check, I say push it and let Michele take a look.

Just remember not to push your system.py changes as we don’t want to overwrite Ymir’s settings ;)

umithardal commented:
@mattclarke Could you please re-review this fundamental implementation? After we decide it is good, I will push it Gerrit as well.

mattclarke commented on nicos_ess/ymir/commands/file_writer.py:
```diff
@@ -0,0 +1,106 @@
+
+    def _monitor_write_process(self):
+        while True:
+            with wait_after(10):  # Validate every ten seconds.
```
As long as the file-writer is writing then we are stuck in this loop, right?
If so, how is the user going to issue a stop_writing command?

umithardal replied:
I have removed the live monitor for the time being as discussed.

mattclarke commented on nicos_ess/devices/datasinks/file_writer.py:
```diff
@@ -0,0 +1,91 @@
+
+
+class FileWriterStatus(KafkaSubscriber, Readable):
+
```
I noticed that most of this class is the same as `KafkaStatusHandler`, so if we inherit from that instead of KafkaSubscriber we only have to override the bits that are different.
This will reduce the amount of code significantly.

umithardal commented:
A new branch has been created with the associated code.
